src/libraries/Microsoft.Extensions.Hosting.WindowsServices/tests/UseWindowsServiceTests.cs

```diff
@@ -2,26 +2,159 @@
+
+            // Remove WindowsServiceLifetime descriptor so we can run this test even when SupportsServiceBase is false.
+            var lifetimeDescriptor = Assert.Single(builder.Services, serviceDescriptor =>
+                serviceDescriptor.ServiceType == typeof(IHostLifetime) &&
+                serviceDescriptor.ImplementationType == typeof(WindowsServiceLifetime));
+            builder.Services.Remove(lifetimeDescriptor);
+
+            using IHost host = builder.Build();
+
+            var eventLogSettings = host.Services.GetRequiredService<IOptions<EventLogSettings>>().Value;
+            Assert.Same(appName, eventLogSettings.SourceName);
+        }
+
+        [ConditionalFact(typeof(UseWindowsServiceTests), nameof(SupportsServiceBase))]
+        public void ServiceCollectionExtensionMethodCanBeCalledOnDefaultConfiguration()
+        {
+            var builder = new HostApplicationBuilder(); 
+
+            // Emulate calling builder.Services.UseWindowsService() from inside a Windows service.
+            UseWindowsServiceUnchecked(builder.Services);
+            // No reason to write event logs in this test.
+            builder.Logging.ClearProviders();
+
+            using IHost host = builder.Build();
+
+            var lifetime = host.Services.GetRequiredService<IHostLifetime>();
+            Assert.IsType<WindowsServiceLifetime>(lifetime);
+        }
+
+
+        [ConditionalFact(typeof(UseWindowsServiceTests), nameof(SupportsServiceBase))]
+        public void ServiceCollectionExtensionMethodThrowsGivenWrongContentRoot()
+        {
+            var builder = new HostApplicationBuilder(new HostApplicationBuilderSettings
+            {
+                ContentRootPath = Path.GetTempPath(),
+            }); 
+
+            // Emulate calling builder.Services.UseWindowsService() from inside a Windows service.
+            UseWindowsServiceUnchecked(builder.Services);
+            // No reason to write event logs in this test.
+            builder.Logging.ClearProviders();
+
+            var ex = Assert.Throws<InvalidOperationException>(() => builder.Build());
+            Assert.Equal($"The IHostEnvironment.ConentRootPath value of '{Path.GetTempPath()}' must match the AppContext.BaseDirectory value of '{AppContext.BaseDirectory}' but is unequal.", ex.Message);
+        }
+
+        private void UseWindowsServiceUnchecked(IServiceCollection services, Action<WindowsServiceLifetimeOptions> configure = null)
+        {
+            _useWindowsServiceUncheckedMethod ??= typeof(WindowsServiceLifetimeHostBuilderExtensions).GetMethod("UseWindowsServiceUnchecked",
+                BindingFlags.Static | BindingFlags.NonPublic, null, new[] { typeof(IServiceCollection), typeof(Action<WindowsServiceLifetimeOptions>) }, null)
+                ?? throw new MissingMethodException();
+
+            configure ??= _ => { };
+            _useWindowsServiceUncheckedMethod.Invoke(null, new object[] { services, configure });
```

Comment:
This is the right pattern for testing non-public APIs rather than InternalsVisibleTo, right?

Reply:
Correct. We opt for reflection over IVT.